src/parquet/arrow/reader.cc

```diff
@@ -1220,6 +1220,68 @@ struct TransferFunctor<::arrow::Decimal128Type, FLBAType> {
   }
 };
 
+/// \brief Convert an arrow::BinaryArray to an arrow::Decimal128Array
+/// We do this by:
+/// 1. Creating an arrow::BinaryArray from the RecordReader's builder
+/// 2. Allocating a buffer for the arrow::Decimal128Array
+/// 3. Converting the big-endian bytes in each BinaryArray entry to two integers
+///    representing the high and low bits of each decimal value.
+template <>
+struct TransferFunctor<::arrow::Decimal128Type, ByteArrayType> {
+  Status operator()(RecordReader* reader, MemoryPool* pool,
+                    const std::shared_ptr<::arrow::DataType>& type,
+                    std::shared_ptr<Array>* out) {
+    DCHECK_EQ(type->id(), ::arrow::Type::DECIMAL);
+
+    // Finish the built data into a temporary array
+    std::shared_ptr<Array> array;
+    RETURN_NOT_OK(reader->builder()->Finish(&array));
+    const auto& binary_array =
+        static_cast<const ::arrow::BinaryArray&>(*array);
+
+    const int64_t length = binary_array.length();
+
+    const auto& decimal_type = static_cast<const ::arrow::Decimal128Type&>(*type);
+    const int64_t type_length = decimal_type.byte_width();
+
+    std::shared_ptr<Buffer> data;
+    RETURN_NOT_OK(::arrow::AllocateBuffer(pool, length * type_length, &data));
+
+    // raw bytes that we can write to
+    uint8_t* out_ptr = data->mutable_data();
+
+    const int64_t null_count = binary_array.null_count();
+
+    // convert each BinaryArray value to valid decimal bytes
+    for (int64_t i = 0; i < length; i++, out_ptr += type_length) {
+
+        int32_t record_len = 0;
+        const uint8_t *record_loc = binary_array.GetValue(i, &record_len);
+
+        if ((record_len < 0) || (record_len > type_length)) {
+            return Status::Invalid("Invalid BYTE_ARRAY size");
+        }
+
+        auto out_ptr_view = reinterpret_cast<uint64_t*>(out_ptr);
+        out_ptr_view[0] = 0;
+        out_ptr_view[1] = 0;
+
+
+        if ((null_count > 0) && !binary_array.IsNull(i)) {
```

Comment:
Nit: The statements in the two `if blocks` are the same. Should we combine them with an `||`?

Reply:
I think I was being too literal in trying to model the structure of  TransferFunctor<::arrow::Decimal128Type, FLBAType>, since that has a for loop when (null_count > 0) and another where it doesn't. 

You are correct, an || is more compact.  I will update.  Thank you.